### PR TITLE
Big integer modular exponentiation (EIP-198)

### PR DESCRIPTION
**Pull request by valiksinev**
Implementation of the big integer modular exponentiation [EIP-198](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-198.md).

_About compute budget costs_
Possible pricing based on 33ns execution time per CU.

[Bench implementation](https://github.com/neonlabsorg/solana/blob/3c1e4f584d34135da5b1a156c83bc83413756815/sdk/benches/bench_big_mod_exp.rs#L92) 

Bellow are results for 32, 64, 96, 128 bytes of the input data respectively:
```
big_mod_exp rnd         time:   [102.10 us 102.22 us 102.37 us] 
big_mod_exp rnd #2      time:   [448.57 us 451.25 us 455.82 us]
big_mod_exp rnd #3      time:   [358.67 us 359.75 us 361.58 us]                               
big_mod_exp rnd #4      time:   [785.48 us 787.26 us 788.92 us]
```                             

32 bytes: 102_000 / 33  = 3090 CU
64 bytes: 451_000 / 33  = 13_666 CU
96 bytes: 359_000 / 33  = 10_878 CU
128 bytes: 787_000 / 33  = 23_848 CU

The execution time depends on the length of the input data. 
With a rough approximation, the dependence can be approximated by the function:

`CU = 180* lenght`

### CODE REVIEW

**Comment by t-nelson** [comment]
@samkim-crypto mind taking this one?  you're probably best qualified on all fronts :slightly_smiling_face:

**Comment by samkim-crypto** [comment]
The exponentiation operations look fine to me. Will probably need another eye on it from our side who works more closely with syscalls, but can you first provide some benches to approximate the compute unit costs similarly to #27961? Updating cargo lock should also fix the CI tests.

**Comment by valiksinev** [comment]
@samkim-crypto 
Your comments were applied.  [See description](https://github.com/solana-labs/solana/pull/28503#issue-1416573694)

File: programs/bpf_loader/src/syscalls/mod.rs [hunk 1]
```diff
@@ -284,6 +287,9 @@ pub fn register_syscalls<'a>(
     // Log data
     syscall_registry.register_syscall_by_name(b"sol_log_data", SyscallLogData::call)?;
 
+    // Big mod exp
+    syscall_registry.register_syscall_by_name(b"sol_big_mod_exp", SyscallBigModExp::call)?;
```
Comment (samkim-crypto) [inline programs/bpf_loader/src/syscalls/mod.rs:291]: Can you register the syscall behind a feature gate? You can use a temporary address, which I can update afterwards with a valid key. You can refer to [this](https://github.com/solana-labs/solana/pull/27961/files#diff-ed9d8e9c9bc044a65944d919be16361b1969bca5122e281ff644a0bb4cbfa198R546) as reference.

File: programs/bpf_loader/src/syscalls/mod.rs [hunk 2]
```diff
@@ -1710,6 +1716,92 @@ declare_syscall!(
...
+            params.modulus_len,
+            invoke_context.get_check_aligned(),
+            invoke_context.get_check_size(),
+        )?;
+
+        let base = BigUint::from_bytes_be(base);
+        let exponent = BigUint::from_bytes_be(exponent);
+        let modulus = BigUint::from_bytes_be(modulus);
+
+        let value = if modulus.is_zero() || modulus.is_one() {
```
Comment (samkim-crypto) [inline programs/bpf_loader/src/syscalls/mod.rs:1782]: Instead of defining the exponentiation operation here, can we call the `big_mod_exp` in the sdk?

**Review by samkim-crypto — DISMISSED** [review]

File: programs/bpf_loader/src/syscalls/mod.rs (hunk 1 above)
Comment (valiksinev) [reply programs/bpf_loader/src/syscalls/mod.rs]: fixed

File: sdk/program/src/big_mod_exp.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        #[derive(serde::Deserialize)]
+        #[serde(rename_all = "PascalCase")]
+        struct TestCase {
+            base: String,
+            exponent: String,
+            modulus: String,
+            expected: String,
+        }
+
+        let test_data = r#"[
```
[the remark this answers is not in the document]
Comment (valiksinev) [reply sdk/program/src/big_mod_exp.rs:74]: fixed

File: programs/bpf_loader/src/syscalls/mod.rs (hunk 2 above)
Comment (valiksinev) [reply programs/bpf_loader/src/syscalls/mod.rs]: fixed

**Comment by samkim-crypto** [comment]
Thanks for all the changes! Looks good to me on the implementation side. Let's just double check everything on the compute costs and then we can finally merge it.

For the benches, we should compute the costs for the worst-case inputs. Instead of computing the benches on random inputs, can we use 
- `[1u8; len]` for the base and exponent
- Use a prime or an RSA modulus for each bit-sizes. You can use something like [this](https://bigprimes.org/RSA-challenge) to generate moduli of appropriate bit sizes

Also, since the syscall could be used for programs that utilizes RSA, can we have benchmarks for up to 2048-bit inputs (32, 64, 128, 256, 512, 1024, 2048)? Testing only up to 128-bit numbers may not be representative of the actual compute cost for very large numbers since `BigUint` represent numbers as vectors of `u64`.

The changes above should not require too much effort given the bench implementation that you linked.

**Comment by valiksinev** [comment]
The bench implementation was updated.
New results:
```
big_mod_exp rsa 32 bits time:     [4.4854 us 4.6145 us 4.7817 us]  
big_mod_exp rsa 64 bits time:     [4.7888 us 4.8973 us 5.0220 us]    
big_mod_exp rsa 128 bits time:    [11.205 us 11.357 us 11.548 us]
big_mod_exp rsa 256 bits time:    [30.799 us 31.114 us 31.473 us]
big_mod_exp rsa 512 bits time:    [140.42 us 142.21 us 144.25 us]                                                                             
big_mod_exp rsa 1024 bits time:   [930.44 us 942.76 us 954.30 us]
big_mod_exp rsa 2048 bits time:   [5.8244 ms 5.8685 ms 5.9160 ms]
```
The execution time depends on the length of the input data.
This dependence is described by a quadratic function:

time = bits^2  (ns)

bits | benchmark result (ns) | bitz^2 
-- | -- | --
32 | 4 614 | 1 024
64 | 4 897 | 4 096
128 | 11 357 | 16 384
256 | 31 473 | 65 536
512 | 142 210 | 262 144
1024 | 942 760 | 1 048 576
2048 | 5 868 500 | 4 194 304

So, the calculation of Compute Units looks like this: 
CU = time / 33 = bits^2 / 33

bits | CU 
-- | -- 
32 | 31
64 | 124 
128 | 496 
256 | 1 986 
  512 | 7 944 
1024 | 31 775 
2048 | 127 100 

But for 2048 bits the CU-price is unacceptably high

**Comment by Lichtso** [comment]
@samkim-crypto You forgot to assign a valid feature ID. Can you do that in a follow up PR?

**Comment by berkantunal** [comment]
When the feature will be activated?

**Comment by Lichtso** [comment]
It is second in the queue for testnet activation.

**Comment by samkim-crypto** [comment]
We are in the process of addressing a security issue regarding this feature, so I moved this feature temporarily to the blocked list in the schedule.

**Comment by guidovranken** [comment]
> We are in the process of addressing a security issue regarding this feature, so I moved this feature temporarily to the blocked list in the schedule.

Can you please reply in the thread where I reported this issue?

**Pull request merged**